src/ocean/meta/values/VisitValue.d

```diff
@@ -0,0 +1,183 @@
+    {
+        /// Visitor can have any internal state it may need
+        long sum;
+
+        /// Returns: 'true' if recursion shall continue for this type
+        bool visit ( T ) ( T* value )
+        {
+            static if (isIntegerType!(T))
+                sum += *value;
+
+            return true;
+        }
+    }
+
+    SumIntegersVisitor visitor;
+    visitValue(some_value, visitor);
+    ---
+
+    Params:
+        value = root value to visit
+        visitor = struct providing templated `visit` method accepting single
+            argument which will be a pointer to visited element
+
+*******************************************************************************/
+
+public void visitValue ( T, Visitor ) ( ref T value, ref Visitor visitor )
+{
+    VisitImpl!(Visitor) impl;
+    impl.visitor = &visitor;
+    impl.visitAll(&value);
+}
+
+/*******************************************************************************
+
+    Implementation of visiting logic. Some types are currently not supported
+    because it is not clear what would recursing mean for them.
+
+    Params:
+        Visitor = user-provided struct implementing templated `visit` method
+            to be called for individual values
+
+*******************************************************************************/
+
+private struct VisitImpl ( Visitor )
+{
+    /// Refers to the persistent instance of visitor struct in case
+    /// implementation / may need internal state
+    Visitor* visitor;
+
+    /***************************************************************************
+
+        Main recursive type/value visiting implementation
+
+        Params:
+            value = pointer to currently processed value. Pointer is used
+                instead of a reference because D1 does not allow defining
+                a reference to a static array
+
+    ***************************************************************************/
+
+    private void visitAll ( T ) ( T* value )
+    {
+        auto deeper = this.visitor.visit(value);
+
+        if (!deeper)
+            return;
+
+        static if (isPrimitiveType!(T))
+        {
+            // do nothing, already processed
+        }
+        else static if (isTypedef!(T))
+        {
+            auto reinterp = cast(TypedefBaseType!(T)*) value;
+            this.visitAll(reinterp);
+        }
+        else static if (isAggregateType!(T))
+        {
+            static if (is(T == class))
+            {
+                bool ignore = *value is null;
+
+                // Recurse into super class fields
+                static if (is(T Bases == super))
+                {
+                    foreach (Base; Bases)
+                    {
+                        static if (Base.init.tupleof.length)
+                            this.visitAll(cast(Base*) value);
+                    }
+                }
+            }
+            else
+            {
+                bool ignore = false;
```

Comment:
These `static if` fake scopes are very deceptive /rant

Reply:
I suggest moving `bool ignore` above `static if` and just setting it to true in the `static if (is(T == class))`, so you avoid fake scope and double definition of the `ignore` and you get rid of `static else` branch.